Miniscript integration follow-ups

darosior opened the pull request:
The Miniscript repository and the Miniscript integration PR here have been a moving target for the past months, and some final cleanups were done there that were not included here. I initially intended to add some small followup commits to #24148 but i think there are enough of them to be worth a followup PR on its own.

Some parts of the code did not change since it was initially written in 2019, and the code could use some modernization. (Use std::optional instead of out args, remove old compiler workarounds).
We refactored the helpers to be more meaningful, and also did some renaming. A new fuzz target was also added and both were merged in a single file. 2 more will be added in #24149 that will be contained in this file too.

The only behaviour change in this PR is to rule out Miniscript with duplicate keys from sane Miniscripts. In a P2WSH context, signatures can be rebounded (Miniscript does not use CODESEPARATOR) and it's reasonable to assume that reusing keys across the Script drops the malleability guarantees.
It was previously assumed such Miniscript would never exist in the first place since a compiler should never create them. We finally agreed that if one were to exist (say, written by hand or from a buggy compiler) it would be very confusing if an imported Miniscript descriptor (after #24148) with duplicate keys was deemed sane (ie, "safe to use") by Bitcoin Core. We now check for duplicate keys in the constructor.

This is (still) joint work with Pieter Wuille. (Actually he entirely authored the cleanups and code modernization.)

darosior committed:
miniscript: introduce a CheckTimeLocksMix helper

This helps to have finer-grained descriptor parsing errors.

darosior committed:
miniscript: remove a workaround for a GCC 4.8 bug

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
miniscript: make equality operator non-recursive

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
miniscript: use optional instead of bool/outarg

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
miniscript: split ValidSatisfactions from IsSane

This makes IsSane clearer. It is useful to differentiate between 'potential non-malleable satisfactions are valid' and 'such satisfactions exist' for testing.

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
miniscript: tiny doc fixups

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
fuzz: rename and improve the Miniscript Script roundtrip target

Parse also key hashes using the Key type. Make this target the first of
the 4 Miniscript fuzz targets in a single `miniscript` file.

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
fuzz: add a Miniscript target for string representation roundtripping

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

darosior committed:
miniscript: mark nodes with duplicate keys as insane

As stated on the website, duplicate keys make it hard to reason about
malleability as a single signature may unlock multiple paths.

We use a custom KeyCompare function instead of operator< to be explicit
about the requirement.

fanquake commented:
[Fuzzer failure](https://github.com/bitcoin/bitcoin/pull/24860/checks?check_run_id=6036314396):
```bash
4.0K	/tmp/cirrus-ci-build/depends/sdk-sources/
4.0K	/tmp/cirrus-ci-build/releases/x86_64-pc-linux-gnu
184 fuzz target(s) found: addition_overflow addr_info_deserialize address_deserialize_v1_notime address_deserialize_v1_withtime address_deserialize_v2 addrman addrman_deserialize addrman_serdeser asmap asmap_direct autofile banman base_encode_decode bech32 block block_deserialize block_file_info_deserialize block_filter_deserialize block_header block_header_and_short_txids_deserialize blockfilter blockheader_deserialize blocklocator_deserialize blockmerkleroot blocktransactions_deserialize blocktransactionsrequest_deserialize blockundo_deserialize bloom_filter bloomfilter_deserialize buffered_file chain checkqueue coins_deserialize coins_view coinselection connman crypto crypto_aes256 crypto_aes256cbc crypto_chacha20 crypto_chacha20_poly1305_aead crypto_common crypto_diff_fuzz_chacha20 crypto_hkdf_hmac_sha256_l32 crypto_poly1305 cuckoocache data_stream_addr_man decode_tx descriptor_parse diskblockindex_deserialize eval_script fee_rate fee_rate_deserialize fees flat_file_pos_deserialize flatfile float golomb_rice hex http_request i2p integer inv_deserialize key key_io key_origin_info_deserialize kitchen_sink load_external_block_file locale merkle_block_deserialize merkleblock message messageheader_deserialize miniscript_script miniscript_string minisketch muhash multiplication_overflow net net_permissions netaddr_deserialize netaddress netbase_dns_lookup node_eviction out_point_deserialize p2p_transport_serialization parse_hd_keypath parse_iso8601 parse_numbers parse_script parse_univalue partial_merkle_tree_deserialize partially_signed_transaction_deserialize policy_estimator policy_estimator_io pow prefilled_transaction_deserialize prevector primitives_transaction process_message process_message_addr process_message_addrv2 process_message_block process_message_blocktxn process_message_cfcheckpt process_message_cfheaders process_message_cfilter process_message_cmpctblock process_message_feefilter process_message_filteradd process_message_filterclear process_message_filterload process_message_getaddr process_message_getblocks process_message_getblocktxn process_message_getcfcheckpt process_message_getcfheaders process_message_getcfilters process_message_getdata process_message_getheaders process_message_headers process_message_inv process_message_mempool process_message_merkleblock process_message_notfound process_message_ping process_message_pong process_message_sendaddrv2 process_message_sendcmpct process_message_sendheaders process_message_tx process_message_verack process_message_version process_message_wtxidrelay process_messages protocol psbt psbt_input_deserialize psbt_output_deserialize pub_key_deserialize random rbf rolling_bloom_filter rpc script script_bitcoin_consensus script_descriptor_cache script_deserialize script_flags script_format script_interpreter script_ops script_sigcache script_sign scriptnum_ops secp256k1_ec_seckey_import_export_der secp256k1_ecdsa_signature_parse_der_lax service_deserialize signature_checker signet snapshotmetadata_deserialize socks5 span spanparsing str_printf string system timedata torcontrol transaction tx_in tx_in_deserialize tx_out tx_pool tx_pool_standard txoutcompressor_deserialize txrequest txundo_deserialize uint160_deserialize uint256_deserialize utxo_snapshot validation_load_mempool versionbits wallet_notifications
184 of 184 detected fuzz target(s) selected: addition_overflow addr_info_deserialize address_deserialize_v1_notime address_deserialize_v1_withtime address_deserialize_v2 addrman addrman_deserialize addrman_serdeser asmap asmap_direct autofile banman base_encode_decode bech32 block block_deserialize block_file_info_deserialize block_filter_deserialize block_header block_header_and_short_txids_deserialize blockfilter blockheader_deserialize blocklocator_deserialize blockmerkleroot blocktransactions_deserialize blocktransactionsrequest_deserialize blockundo_deserialize bloom_filter bloomfilter_deserialize buffered_file chain checkqueue coins_deserialize coins_view coinselection connman crypto crypto_aes256 crypto_aes256cbc crypto_chacha20 crypto_chacha20_poly1305_aead crypto_common crypto_diff_fuzz_chacha20 crypto_hkdf_hmac_sha256_l32 crypto_poly1305 cuckoocache data_stream_addr_man decode_tx descriptor_parse diskblockindex_deserialize eval_script fee_rate fee_rate_deserialize fees flat_file_pos_deserialize flatfile float golomb_rice hex http_request i2p integer inv_deserialize key key_io key_origin_info_deserialize kitchen_sink load_external_block_file locale merkle_block_deserialize merkleblock message messageheader_deserialize miniscript_script miniscript_string minisketch muhash multiplication_overflow net net_permissions netaddr_deserialize netaddress netbase_dns_lookup node_eviction out_point_deserialize p2p_transport_serialization parse_hd_keypath parse_iso8601 parse_numbers parse_script parse_univalue partial_merkle_tree_deserialize partially_signed_transaction_deserialize policy_estimator policy_estimator_io pow prefilled_transaction_deserialize prevector primitives_transaction process_message process_message_addr process_message_addrv2 process_message_block process_message_blocktxn process_message_cfcheckpt process_message_cfheaders process_message_cfilter process_message_cmpctblock process_message_feefilter process_message_filteradd process_message_filterclear process_message_filterload process_message_getaddr process_message_getblocks process_message_getblocktxn process_message_getcfcheckpt process_message_getcfheaders process_message_getcfilters process_message_getdata process_message_getheaders process_message_headers process_message_inv process_message_mempool process_message_merkleblock process_message_notfound process_message_ping process_message_pong process_message_sendaddrv2 process_message_sendcmpct process_message_sendheaders process_message_tx process_message_verack process_message_version process_message_wtxidrelay process_messages protocol psbt psbt_input_deserialize psbt_output_deserialize pub_key_deserialize random rbf rolling_bloom_filter rpc script script_bitcoin_consensus script_descriptor_cache script_deserialize script_flags script_format script_interpreter script_ops script_sigcache script_sign scriptnum_ops secp256k1_ec_seckey_import_export_der secp256k1_ecdsa_signature_parse_der_lax service_deserialize signature_checker signet snapshotmetadata_deserialize socks5 span spanparsing str_printf string system timedata torcontrol transaction tx_in tx_in_deserialize tx_out tx_pool tx_pool_standard txoutcompressor_deserialize txrequest txundo_deserialize uint160_deserialize uint256_deserialize utxo_snapshot validation_load_mempool versionbits wallet_notifications
Fuzzing harnesses lacking a corpus: coinselection miniscript_script miniscript_string wallet_notifications
Please consider adding a fuzz corpus at https://github.com/bitcoin-core/qa-assets
Run address_deserialize_v2 with args ['/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz', '-runs=1', '/tmp/cirrus-ci-build/ci/scratch/qa-assets/fuzz_seed_corpus/address_deserialize_v2']fuzz: key.cpp:392: void ECC_Start(): Assertion `secp256k1_context_sign == nullptr' failed.

fuzz: key.cpp:392: void ECC_Start(): Assertion `secp256k1_context_sign == nullptr' failed.

Target "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz -runs=1 /tmp/cirrus-ci-build/ci/scratch/qa-assets/fuzz_seed_corpus/address_deserialize_v2" failed with exit code -6
```

Sjors commented:
Did some very light code review, it looks sane. I didn't look at the fuzz changes.

Disallowing duplicate keys seems reasonable; I would not at all be surprised if people generate miniscript by hand or even from Sipa's website, and then paste whatever keys they need into the result.

Nit: can you make b9a36d3c6ce77c98e5a6e76d0ed223c410f323d2 a scripted diff?

darosior commented:
Thanks for the notification about the fuzzer failure, i fixed it. See also discussion about it in https://github.com/sipa/miniscript/pull/106.

Regarding the scripted diff, sure. I'll push it soon.

darosior committed:
scripted-diff: miniscript: rename 'nodetype' variables to 'fragment'

The 'Fragment' type was previously named 'Nodetype'. For clarity, name
the variables the same.

-BEGIN VERIFY SCRIPT-
sed -i 's/nodetype/fragment/g' src/script/miniscript.*
-END VERIFY SCRIPT-

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

vincenzopalazzo commented on src/script/miniscript.cpp:
```diff
@@ -282,16 +280,15 @@ size_t ComputeScriptLen(Fragment nodetype, Type sub0typ, size_t subsize, uint32_
 }
 
-bool DecomposeScript(const CScript& script, std::vector<std::pair<opcodetype, std::vector<unsigned char>>>& out)
+std::optional<std::vector<std::pair<opcodetype, std::vector<unsigned char>>>> DecomposeScript(const CScript& script)
```
I'm not a fan of typedef, but I think that one typedef, in this case, can make the return type cleaner from a reader's viewpoint?

```suggestion
// FIXME: move this declaration in some better place!
typedef std::pair<opcodetype, std::vector<unsigned char>> OPCode

std::optional<std::vector<OPCode>> DecomposeScript(const CScript& script)
```

darosior replied:
Done

vincenzopalazzo reviewed (commented):
Concept ACK, with some opinionable comments!

darosior commented:
cc @sanket1729

sanket1729 commented on src/script/miniscript.h:
```diff
@@ -779,8 +779,11 @@ struct Node {
 
-    //! Do all sanity checks.
-    bool IsSane() const { return IsValid() && IsNonMalleable() && CheckTimeLocksMix() && CheckOpsLimit() && CheckStackSize(); }
+    //! Whether successful non-malleable satisfactions are guaranteed to be valid.
```
I think you mean all satisfactions here? Not only the non-malleable ones?

sipa replied:
Only the non-malleable ones, because CheckOpsLimit and CheckStackSize only take the logic of the non-malleable satisfaction algorithm into account.

sanket1729 commented on src/script/miniscript.h:
```diff
@@ -779,8 +779,11 @@ struct Node {
+    bool ValidSatisfactions() const { return IsValid() && CheckOpsLimit() && CheckStackSize(); }
+
+    //! Whether the apparent policy of this node matches its script semantics.
+    bool IsSane() const { return ValidSatisfactions() && IsNonMalleable() && CheckTimeLocksMix(); }
```
Technically, all timelocks are malleable if there is no signature. If there is no signature, anyone can malleate nLockTime/nSequence to any value perhaps not matching the script semantics?  

Why is the NeedsSignature check only on top-level? and not a general condition for sanity?

darosior replied:
You are right, `n:older(12)` shouldn't be called "sane". But to call it "sane" one should use `IsSaneTopLevel()`.

I think that the check for a fragment's sanity should be context-less, and the requirement for signature depends on the context. To take back the example from above `n:older(12)` shouldn't be called sane on its own, but it should in the context of `and_b(n:older(12),s:pk(A))`. So it should only be called insane if not 's' and is the top level fragment. That's what `IsSaneTopLevel()` is.
It's not purely theoretical, for instance we use this in #24148 to find the deepest insane sub of an insane Miniscript: https://github.com/bitcoin/bitcoin/pull/24148/files#diff-a55760aaec4bce216663f5ebf65823516347356a8320d30459427149f7bbc2c5R824-R831. If the `IsSane()` check is context-dependent it gets messy.

Now, maybe there is a point for the need for a better naming.

darosior replied:
Done.

sanket1729 commented on src/script/miniscript.h:
```diff
@@ -746,6 +762,30 @@ struct Node {
+    //! Returns true if the node contains at least one duplicate key.
+    bool ContainsDuplicateKey() const {
+        auto upfn = [this](const Node& node, Span<std::set<Key>> subs) -> std::optional<std::set<Key>> {
+            if (&node != this && node.duplicate_key) return {};
```
I am confused by why we need the node = this here. Should upfn always not operate on this node?

sipa replied:
`this->duplicate_key` isn't yet initialized when `ContainsDuplicateKey` is being called, so we need to skip inspecting it.

sanket1729 reviewed (approved):
LGTM. Few questions.

darosior committed:
miniscript: add an OpCode typedef for readability

Suggested-by: Vincenzo Palazzo

darosior committed:
miniscript: explicit the threshold size computation in multi()

darosior committed:
miniscript: nit: don't return after assert(false)

darosior commented:
I think i addressed all comments here. It's ready for another round of review. :)

I had previously made the rename a scripted diff as asked by @Sjors. In the last push i:
- Added a `typedef` for the (operand, data pushed) pair as suggested by @vincenzopalazzo 
- Added a `CheckDuplicateKey()` which is useful for unit tests and for granular error detection in #24148 
- Addressed the nits from @sanket1729 review of #24148

darosior committed:
miniscript: rename IsSane and IsSaneSubexpression to prevent misuse

darosior commented:
Added an explicit requirement on the context to provide a `KeyCompare` method instead of implicitly relying on `operator<` for checking key duplicates. Suggested by @sipa in https://github.com/bitcoin/bitcoin/pull/24148#issuecomment-1127664730.

sipa commented on src/script/miniscript.h:
```diff
@@ -746,6 +763,36 @@ struct Node {
+     * for more complicated types.
+     */
+    template<typename Ctx> bool ContainsDuplicateKey(const Ctx& ctx) const {
+        using set = std::set<Key, std::function<bool(const Key& a, const Key& b)>>;
```
`std::function` may have allocation overhead, and results in dynamic binding (functions pointers rather than compile-time resolving). Here is an alternative, though it's not as elegant as I'd like:

```c++
--- a/src/script/miniscript.h
+++ b/src/script/miniscript.h
@@ -767,14 +767,20 @@ public:
      * for more complicated types.
      */
     template<typename Ctx> bool ContainsDuplicateKey(const Ctx& ctx) const {
-        using set = std::set<Key, std::function<bool(const Key& a, const Key& b)>>;
+        // We cannot use a lambda here, as lambdas are non assignable, and the set operations
+        // below require moving the comparators around.
+        struct Comp {
+            const Ctx* ctx_ptr;
+            Comp(const Ctx& ctx) : ctx_ptr(&ctx) {}
+            bool operator()(const Key& a, const Key& b) const { return ctx_ptr->KeyCompare(a, b); }
+        };
+        using set = std::set<Key, Comp>;
 
         auto upfn = [this, &ctx](const Node& node, Span<set> subs) -> std::optional<set> {
             if (&node != this && node.duplicate_key) return {};
 
             size_t keys_count = node.keys.size();
-            auto comp = [&ctx](const Key& a, const Key& b) -> bool { return ctx.KeyCompare(a, b); };
-            set key_set{node.keys.begin(), node.keys.end(), comp};
+            set key_set{node.keys.begin(), node.keys.end(), Comp(ctx)};
             if (key_set.size() != keys_count) return {};
 
             for (auto& sub: subs) {
```

darosior replied:
[TIL](https://stackoverflow.com/a/9088690/12967587), thanks. Done.

sipa commented:
utACK f552056225f2205bcb4af875e42730a357ceb916 (with the caveat that a lot of it is my own code).

With or without addressing https://github.com/bitcoin/bitcoin/pull/24860#discussion_r883686621

sipa commented:
utACK f3a50c9dfe645c548713e44e0eaf26ea9917a379 (with the caveat that a lot of it is my own code).

sanket1729 reviewed (approved):
code review ACK f3a50c9dfe645c548713e44e0eaf26ea9917a379. Did not review the fuzz tests.

MarcoFalke commented:
Renamed the fuzz target input folder to preserve the inputs: https://github.com/bitcoin-core/qa-assets/commit/3366f7bad1144804e372afd88ee3eae6ca140dfa

darosior commented:
I've been populating corpora for the two targets here. I'll PR them to qa-assets soon.
------- Original Message -------
Le mercredi 8 juin 2022 à 8:09 PM, MacroFake ***@***.***> a écrit :

> Renamed the fuzz target input folder to preserve the inputs: ***@***.***(https://github.com/bitcoin-core/qa-assets/commit/3366f7bad1144804e372afd88ee3eae6ca140dfa)
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/bitcoin/bitcoin/pull/24860#issuecomment-1150233735), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AFLK3F2HMLUCLFEVIE3XPQLVODOUHANCNFSM5TP67B6Q).
> You are receiving this because you authored the thread.Message ID: ***@***.***>

MarcoFalke commented on src/test/fuzz/miniscript.cpp:
```diff
@@ -0,0 +1,167 @@
+{
+    FuzzedDataProvider provider(buffer.data(), buffer.size());
+    auto str = provider.ConsumeRemainingBytesAsString();
+    auto parsed = miniscript::FromString(str, PARSER_CTX);
```
Looks like this may take quite a while. For example:

[clusterfuzz-testcase-miniscript_string-5717444439703552.bin.txt](https://github.com/bitcoin/bitcoin/files/8867405/clusterfuzz-testcase-miniscript_string-5717444439703552.bin.txt)

takes 10 seconds

With the flame graph:

![Screenshot from 2022-06-09 08-14-22](https://user-images.githubusercontent.com/6399679/172777704-a2110ad7-dbb9-4469-a94b-08767745db32.png)

darosior replied:
Yeah calling it in the constructor was not exactly bright since we'll re-do the same computation as we pile more nodes in the parsers. On the other hand it's hard not to do so and to keep the `const`ness of the `NodeRef`. Here are different approaches i tried:
- Cache the duplicate key result only once we call it https://github.com/darosior/bitcoin/tree/miniscript_cache_dup_key_check. Needs to drop the constness.
- Conditionally skip the check in the constructor, in order to only do it for the top level node https://github.com/darosior/bitcoin/tree/miniscript_skip_dup_key_check. Doesn't drop the constness but not elegant..
- Don't cache it at all https://github.com/darosior/bitcoin/tree/miniscript_no_dup_key_cache

sipa replied:
@darosior What about this idea:

* Make the current `miniscript::Node` constructors private, and make them not compute the `duplicate_key` value.
* Add a forwarding public constructor which invokes the private ones, and also computes the `duplicate_key` value (so now any publicly constructed Node object will have duplicate_key set, but it's possible to bypass this computation by using the private ones, if you have access to them).
* `FromScript` and `FromString` are made `friend`, and thus can use the private constructors.
* Make the `duplicate_key` variable mutable, so `ContainsDuplicateKey` can modify it (and is changed so it sets the value for all children of the called node too).

darosior replied:
Sounds much better than all my approaches. I'll implement and PR this.

MarcoFalke replied:
See also https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=47994
